feat: added hcp consul docs (#16083)
newrelic/docs-website: src/content/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/collector/collector-configuration-examples/opentelemetry-collector-hcp-consul.mdx

---
title: Collector for monitoring HCP Consul metrics
tags:
  - Integrations
  - Open source telemetry integrations
  - OpenTelemetry
  - Consul
  - HCP
  - Hashicorp
  - Networking
  - Server
  - Envoy
metaDescription: You can collect metrics from HCP managed Consul using the OpenTelemetry Collector.

> freshnessValidatedDate: never

Reviewer: ```suggestion
freshnessValidatedDate: 2024-02-26
```

---

> You can collect metrics about your HCP-managed Consul deployment with the OpenTelemetry Collector. The collector is a component of OpenTelemetry that collects, processes, and exports telemetry data to New Relic (or any observability backend).
> 
> This example works by running a prometheus receiver configuration inside the OpenTelemetry collector, which scrapes [HCP Consul's service metrics API](https://developer.hashicorp.com/hcp/docs/consul/monitor/metrics) and exports that data to [New Relic's OpenTelemetry API endpoint](https://docs.newrelic.com/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/get-started/opentelemetry-set-up-your-app/#note-endpoints).
> 
> HCP Consul emits two forms of observability metrics, [Service metrics](https://developer.hashicorp.com/hcp/docs/consul/monitor/consul-central/observability#server-metrics) and [Envoy proxy metrics](https://developer.hashicorp.com/hcp/docs/consul/monitor/consul-central/observability#envoy-proxy-metrics). The example collector used in this guide will get you started with HCP Consul server metrics. For Envoy metrics, you can reference the notes below on Envoy metrics or check the [HCP Documentation](https://developer.hashicorp.com/hcp/docs/consul/monitor/consul-central/observability/telemetry-collector#forward-metrics-to-another-collector).
> 
> Complete the steps below to collect metrics from HCP consul and export them to New Relic.

Reviewer: ```suggestion
This guide explains how to collect metrics from your HCP-managed Consul deployment using the OpenTelemetry Collector. The collector gathers, processes, and sends telemetry data (including metrics, logs, and traces) to various observability backends, including New Relic.

The setup example below leverages the Prometheus receiver within the collector. It retrieves service metrics directly from [HCP Consul's service metrics API](https://developer.hashicorp.com/hcp/docs/consul/monitor/metrics) and exports them to New Relic's OpenTelemetry endpoint.

HCP Consul offers two types of observability metrics:
* [Server metrics](https://developer.hashicorp.com/hcp/docs/consul/monitor/consul-central/observability#server-metrics): These track Consul server performance and resource usage. 
* [Envoy proxy metrics](https://developer.hashicorp.com/hcp/docs/consul/monitor/consul-central/observability#envoy-proxy-metrics): These monitor the Envoy service mesh within Consul. 

Follow the steps below to collect and export HCP Consul metrics to New Relic.
```

<Steps>
    <Step>

>         ## Make sure you're set up

Reviewer: ```suggestion
        ## Review the requirements [#requirements]
```

>       Before you start, you need to have the <InlinePopover type="licenseKey" /> for the account you want to report data to. You should also verify that:
> 
>       * You have a docker daemon running
>       * You have [Docker Compose](https://docs.docker.com/compose/) installed
>       * You have your [HCP Consul ACL admin token](https://developer.hashicorp.com/hcp/docs/consul/hcp-managed/access#generate-admin-token) available 

Reviewer: ```suggestion
      Before you start, you'll need the following:
     
      * A running Docker daemon service
      * An installed [Docker Compose](https://docs.docker.com/compose/) tool
      * Your [HCP Consul ACL admin token](https://developer.hashicorp.com/hcp/docs/consul/hcp-managed/access#generate-admin-token) 
      * Your New Relic <InlinePopover type="licenseKey" /> for the account you want to report data to
```

    </Step>

    <Step>

>         ## Download or clone the example repos

Reviewer: ```suggestion
        ## Download or clone the example configuration [#download]
```

>         Download [New Relic's OpenTelemetry Examples repo](https://github.com/newrelic/newrelic-opentelemetry-examples) as this setup uses its example collector configuration. Once installed, open the [HCP Consul example](https://github.com/newrelic/newrelic-opentelemetry-examples/tree/main/other-examples/collector/hcp-consul) directory. For more information, you can check the `README` there as well.

Reviewer: ```suggestion
        1. [Download the example Configuration](https://github.com/newrelic/newrelic-opentelemetry-examples).
        2. Once downloaded, locate the directory named `hcp-consul` within the `other-examples/collector` folder of the repository. This directory contains the specific configuration for collecting HCP Consul metrics.
        3. (Optional) Refer to the [README](https://github.com/newrelic/newrelic-opentelemetry-examples/tree/main/other-examples/collector/hcp-consul#readme) for additional details and instructions related to the HCP Consul example.
```

    </Step>

    <Step>

>       ##  Set environment variables and run the collector
>         * Set the HCP Access Token and Access URL values in the `.env` file
>         * Set the New Relic API key variable to your New Relic license ingest key in the `.env` file

Reviewer: ```suggestion
      ##  Set environment variables and run the collector [#set-variables-run-collector]
      
      In the `.env` file:
        * Set the HCP Access Token and Access URL values
        * Set the New Relic API key variable to your New Relic license key
```

```bash
  # Open the HCP Consul example directory
  cd newrelic-opentelemetry-examples/other-examples/collector/hcp-consul

>   # Set environment variables.

Author: ```suggestion
```

  # run the collector in docker

>   docker compose up

Reviewer: @jcountsNR  Can you add to this bash code block an example of the environment variables (mentioned in the bullet points above)?

Author: The environment variables are suggested to be set using the `.env` file in this case, which has comments and links to docs. It wouldn't be set from a bash command. It might make sense to just remove the comment from the command since it's specified above.

```

>       ### Local Variable information

Reviewer: ```suggestion
      ### Environment variable information
      
      If you need more information on the environment variables set above, see the documentation in the following table.
```

        <table>
          <thead>
            <tr>
              <th style={{ width: "200px"}}>
                Variable
              </th>
              <th>
                Description
              </th>
              <th>
                Docs
              </th>
            </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  `NEW_RELIC_API_KEY`
                </td>
                <td>
                  New Relic Ingest API Key
                </td>
                <td>

>                   [API Key docs](/docs/apis/intro-apis/new-relic-api-keys/)

Reviewer: ```suggestion
                  [Learn about API keys](/docs/apis/intro-apis/new-relic-api-keys/)
```

                </td>
              </tr>
              <tr>
                <td>
                  `NEW_RELIC_OTLP_ENDPOINT`
                </td>
                <td>

>                   Default US New Relic OTLP endpoint is https://otlp.nr-data.net:4318

Reviewer: ```suggestion
                  The default US New Relic OTLP endpoint is `https://otlp.nr-data.net:4318`.
```

                </td>
                <td>

>                   [OTLP endpoint config docs](/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/get-started/opentelemetry-set-up-your-app/#review-settings)

Reviewer: ```suggestion
                  [Review OTLP endpoint configurations](/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/get-started/opentelemetry-set-up-your-app/#review-settings)
```

                </td>
              </tr>
              <tr>
                <td>
                  `HCP_ACCESS_TOKEN`
                </td>
                <td>
                  ACL Admin token for HCP Consul
                </td>
                <td>

>                   [Docs for creating an ACL admin token](https://developer.hashicorp.com/hcp/docs/consul/hcp-managed/access#generate-admin-token)

Reviewer: ```suggestion
                  [Create an ACL admin token](https://developer.hashicorp.com/hcp/docs/consul/hcp-managed/access#generate-admin-token)
```

                </td>
              </tr>
              <tr>
                <td>
                  `HCP_ACCESS_URL`
                </td>
                <td>
                  URL to access your specific Consul server
                </td>
                <td>

>                   [docs to locate your access URL](https://developer.hashicorp.com/hcp/docs/consul/hcp-managed/access#get-access-url)

Reviewer: ```suggestion
                  [Locate your access URL](https://developer.hashicorp.com/hcp/docs/consul/hcp-managed/access#get-access-url)
```

                </td>
              </tr>
            </tbody>
        </table>

    </Step>
    <Step>

>       ## HCP Consul Envoy Metrics
>         To monitor HCP Consul Envoy metrics in New Relic, you can configure the OpenTelemetry collector built into your HCP Consul cluster to forward metrics to your own collector. This can include the collector running in this example.
> 
> 
>         First you will need to add an http reciever to the OpenTelemetry collector. you can add this into the `receivers` group in the example `collector.yaml` file.
> 

Reviewer: ```suggestion
      ## Forward HCP Consul Envoy metrics to New Relic [#forward-envoy-metrics]
      
        To monitor HCP Consul Envoy metrics in New Relic, you can configure the OpenTelemetry collector built into your HCP Consul cluster to forward metrics to your own collector. This can include the collector running in this example.

        1. Add an HTTP receiver to the OpenTelemetry collector. You can add this into the `receivers` group in the example `collector.yaml` file:

```

          ```yaml
            receivers:
              otlp:
                protocols:
                  http:
          ```

>         This will add a receiver listening at `0.0.0.0:4318`. Once that is set up, you can then configure your HCP Consul cluster to forward envoy metrics to this receiver. This can be configured from the `telemetryCollector` stanza in your HCP Consul configuration file. Here is an example configuration you can use:

Reviewer: ```suggestion
        This will add a receiver listening at `0.0.0.0:4318`. 
        
        2. Configure your HCP Consul cluster to forward Envoy metrics to this receiver. This can be configured from the `telemetryCollector` stanza in your HCP Consul configuration file. For example:
```

          ```yaml
              telemetryCollector:
                enabled: true
                customExporterConfig: |-
                  {
                    "exporter_config": {
                      "otlphttp": {
                        "endpoint": "0.0.0.0:4318",
                        "headers": {
                          "authorization": "<auth>"
                        },
                        "timeout": "2s"
                      }
                    }
                  }

          ```

    </Step>
    <Step>

>        ## View your data in New Relic

Reviewer: ```suggestion
       ## View your data in New Relic [#view-data]
```

>       You can view your HCP Consul server data in a few different ways.

Reviewer: ```suggestion
      In New Relic, you can view your HCP Consul data in a few different places:
```

>       * Navigate to the [New Relic marketplace](https://one.newrelic.com/marketplace) and search for `consul`. The available dashboards can be installed right onto your account!

Reviewer: ```suggestion
      * Navigate to the [New Relic marketplace](https://one.newrelic.com/marketplace) and search for "consul". Click one of the dashboards to quickly install it for your account.
```

>       * Navigate to the metrics explorer and filter for `consul`, or `envoy` for envoy metrics. This data can be added to any custom alert or dashboard.      

Reviewer: ```suggestion
      * Navigate to the metrics explorer and filter for `consul`, or `envoy` for Envoy metrics. This data can be added to any custom alert or dashboard.      
```
